src/lib/cache/util.go

```diff
@@ -0,0 +1,73 @@
+// Copyright Project Harbor Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//    http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package cache
+
+import (
+	"errors"
+	"reflect"
+	"sync"
+)
+
+func indirect(reflectValue reflect.Value) reflect.Value {
+	for reflectValue.Kind() == reflect.Ptr {
+		reflectValue = reflectValue.Elem()
+	}
+	return reflectValue
+}
+
+func simpleCopy(dst interface{}, src interface{}) error {
+	from := indirect(reflect.ValueOf(src))
+	to := indirect(reflect.ValueOf(dst))
+	if !to.CanAddr() {
+		return errors.New("dst value is unaddressable")
+	}
+
+	if !from.Type().ConvertibleTo(to.Type()) {
+		return errors.New("src value is not convertible to the dst value")
+	}
+
+	to.Set(from.Convert(to.Type()))
+
+	return nil
+}
+
+type keyMutex struct {
+	m *sync.Map
+}
+
+func (km keyMutex) Lock(key interface{}) {
+	m := sync.Mutex{}
+	act, _ := km.m.LoadOrStore(key, &m)
+
+	mm := act.(*sync.Mutex)
+	mm.Lock()
+	if mm != &m {
```

Comment:
check the second ret value of `LoadOrStore` and you don't need this `if`?

Reply:
May because that `mm != &m` is clear to understand and will never be a mistake.

This if statement checks that the mutex load from the `sync.Map` must be the mutex created in the `Lock` function.  When it's not the mutex created in the `Lock` function, it means that another goroutine locked the key before this current goroutine, so the current goroutine needs to `Lock` the key again.

Reply:
I'm a little confused.... 
You did not remove `mm` from the map, so in the next `Lock` it's still `mm != &m` ?

Reply:
I misunderstood, the point is that a go-routine needs to create a new mutex and manage the whole lifecycle of the mutex until the key is released.